DDP-8543: DSM Local Environment Support

bskinner opened the pull request:
DDP-8543

## Context

Enables use of a locally hosted services to run DSM.

Changes in this PR include
* Adds a `pubsub.emulator` boolean key to the configuration. Set this to `true` if using a [Pub/Sub Emulator](https://cloud.google.com/pubsub/docs/emulator). This value is optional, and defaults to `false`.
* Adds a `pubsub.host` string key to the configuration. This should be set to the authority (`host:port`) of the Pub/Sub Emulator host.

## Testing
See the [Running a localhost DSM](https://broadinstitute.atlassian.net/wiki/spaces/DDP/pages/3255107618/Running+a+localhost+DSM) document on the Broad wiki for instructions on how to make use of these changes, and for several example files.

bskinner committed:
Initial PubSub Emulator work for running locally

bskinner committed:
Adds a local docker support environment

bskinner committed:
Allow for a null/blank password or username

bskinner committed:
Adds example run configurations for VSCode & IDEA

bskinner committed:
Link to ancillary docs in main readme

bskinner committed:
Pub/Sub changes for emulator use

bskinner committed:
Adds an example configuration for DSM

bskinner committed:
Should be the topic name

bskinner committed:
Update the default config paths

bskinner committed:
Configuration fixes

bskinner committed:
Enable Pub/Sub Emulator use for all Subscribers

bskinner committed:
Logging doesn't need to be as verbose out of the gate

bskinner commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/util/ElasticsearchServiceUtil.java:
```diff
@@ -61,12 +61,17 @@ public static Map<ElasticSearchIndexType, String> getIndicesForStudy(
+                    new UsernamePasswordCredentials(userName, password));
+        } else {
+            credentialsProvider = null;
+        }
```
Enables use of the local Elasticsearch instance as the provided setup does not require auth.

bskinner commented on pepper-apis/dsm-server/src/resources/config/application.conf.default:
```diff
@@ -0,0 +1,251 @@
+//
+GoogleAnalytics.trackingId = "google-analytics-token"
+
+projectId =  "google-project-id"
```
There are at least 4 Google Project ID keys:
* `projectId`
* `google.secret.project`
* `portal.googleProjectName`
* `pubsub.projectId`

In practice, all of them should be set to the same value.

bskinner commented on pepper-apis/local-support-environment/docker-compose.yml:
```diff
@@ -0,0 +1,67 @@
+    image: docker.elastic.co/kibana/kibana:8.3.3
+    environment:
+      ELASTICSEARCH_HOSTS: '["http://elasticsearch:9200"]'
+      XPACK_ENCRYPTEDSAVEDOBJECTS_ENCRYPTIONKEY: 0123456789abcdeffedcba9876543210
+      XPACK_REPORTING_ENCRYPTIONKEY: 0123456789abcdeffedcba9876543210
+      XPACK_SECURITY_ENCRYPTIONKEY: 0123456789abcdeffedcba9876543210
```
These values are for **local use only**, and are not considered to be "secret".

bskinner commented on pepper-apis/local-support-environment/docker-compose.yml:
```diff
@@ -0,0 +1,67 @@
+      MYSQL_USER: pepper
+      MYSQL_PASSWORD: pepper-dev
+      MYSQL_ROOT_PASSWORD: pepper-dev
+      TZ: "America/New_York"
```
This should be the UTC timezone
```suggestion
      TZ: "Etc/UTC"
--


[Etc/UTC](https://en.wikipedia.org/wiki/Coordinated_Universal_Time)"
```

bskinner commented on pepper-apis/dsm-server/src/resources/config/application.conf.default:
```diff
@@ -0,0 +1,251 @@
+  host = "127.0.0.1:8085"
+
+  topic = "dsx-local-dsm-test-results"
+  subscription = "dsx-local-dsm-test-results-sub"
+
+  dsm_to_dss_topic = "dsx-local-dss-tasks"
+  dss_tasks_topic = "dsx-local-dss-tasks"
+
+  dss_to_dsm_topic = "dsx-local-dss-task-results"
+  dss_to_dsm_subscription = "dsx-local-dss-task-results-sub"
+
+  dsm_tasks_topic = "dsx-local-dsm_tasks_topic"
+  dsm_tasks_subscription = "dsx-local-dsm-tasks-sub"
+
+  dsm_to_mercury_topic = "dsx-local-mercury-pepper"
```
Prefixed with `dsx-local` to make it easy to identify topics/subscriptions if they are accidentally created outside of a localhost environment.

bskinner committed:
Usage clarification

bskinner committed:
Excessive whitespace

bskinner commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/DSMServer.java:
```diff
@@ -706,11 +767,30 @@ private void setupPubSub(@NonNull Config cfg, NotificationUtil notificationUtil)
+                .setExecutorProvider(resultsSubExecProvider)
+                .setMaxAckExtensionPeriod(org.threeten.bp.Duration.ofSeconds(120));
+
+            if (!StringUtils.isBlank(emulatorHost)) {
```
`isBlank` works, but `isEmpty` feels like a better option, behavioral-wise (that is, `pubsub.host = "   "` would result in a URI format option, instead of a `null` return here)

bskinner committed:
Set MySQL Docker TZ to UTC

bskinner commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/DSMtasksSubscription.java:
```diff
@@ -92,11 +103,30 @@ public static void subscribeDSMtasks(String projectId, String subscriptionId) {
-        subscriber = Subscriber.newBuilder(resultSubName, receiver).setParallelPullCount(1).setExecutorProvider(resultsSubExecProvider)
-                .setMaxAckExtensionPeriod(org.threeten.bp.Duration.ofSeconds(120)).build();
+
+        var subscriberBuilder  = Subscriber.newBuilder(resultSubName, receiver)
+                .setParallelPullCount(1)
+                .setExecutorProvider(resultsSubExecProvider)
+                .setMaxAckExtensionPeriod(org.threeten.bp.Duration.ofSeconds(120));
+        
+        if (!StringUtils.isBlank(pubSubHost)) {
+            var channel = ManagedChannelBuilder
+                .forTarget(pubSubHost)
+                .usePlaintext()
+                .build();
+            
+            subscriberBuilder.setChannelProvider(FixedTransportChannelProvider.create(GrpcTransportChannel.create(channel)));
+        }
+
+        if (emulator) {
+            subscriberBuilder.setCredentialsProvider(NoCredentialsProvider.create());
+        }
+
+        var subscriber = subscriberBuilder.build();
```
This setup code for subscribers appears in multiple spots and is a prime candidate for a DRY refactoring, but I don't see a clear spot for the code to live. I'm open to suggestions.

bskinner commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/util/ElasticsearchServiceUtil.java:
```diff
@@ -61,12 +61,17 @@ public static Map<ElasticSearchIndexType, String> getIndicesForStudy(
-
-        final CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
-        credentialsProvider.setCredentials(AuthScope.ANY,
-                new UsernamePasswordCredentials(userName, password));
+        String userName = ConfigUtil.getStrIfPresent(cfg, ConfigFile.ELASTICSEARCH_USERNAME);
+        String password = ConfigUtil.getStrIfPresent(cfg, ConfigFile.ELASTICSEARCH_PASSWORD);
```
This is in the `dss-core` package- removing it!

bskinner committed:
Enables use of a local ES instance with auth disabled

bskinner committed:
Revert "Allow for a null/blank password or username"

The original change was in `dss-core` and not `dsm-core`.

This reverts commit bb2514a86055959f06791c4bf173421fa07afbaf.

bskinner committed:
Checkstyle

bskinner committed:
Relocate the example config to the proper resources dir

bskinner committed:
Comments for the GA token config

bskinner committed:
Copy-paste shrapnel

bskinner committed:
Never accessed- not necessary for operation!

devonbush reviewed (commented):
This looks awesome!  Has anyone gotten this working on an M1 mac?  When I try I get 
```
~/code/ddp/ddp-study-server/pepper-apis/local-support-environment (dsm-getting-started)$ docker compose up --build
[+] Running 0/4
 ⠿ redis Error                                                                                                                                                                                       1.1s
 ⠿ kibana Error                                                                                                                                                                                      1.1s
 ⠙ db Pulling                                                                                                                                                                                        1.1s
 ⠿ elasticsearch Error                                                                                                                                                                               1.1s
no matching manifest for linux/arm64/v8 in the manifest list entries
```

trying again with a Rosetta terminal fixes redis, but not elasticSearch or Kibana.
```
~/code/ddp/ddp-study-server/pepper-apis/local-support-environment (dsm-getting-started)$ docker compose up --build
[+] Running 0/10
 ⠿ elasticsearch Error                                                                                                                                                                               0.8s
 ⠏ db Pulling                                                                                                                                                                                        0.8s
 ⠏ redis Pulling                                                                                                                                                                                     0.8s
   ⠹ f97344484467 Pulling fs layer                                                                                                                                                                   0.2s
   ⠹ 491979375772 Pulling fs layer                                                                                                                                                                   0.2s
   ⠹ fa2ce6183da4 Pulling fs layer                                                                                                                                                                   0.2s
   ⠹ dc0978d3f333 Waiting                                                                                                                                                                            0.2s
   ⠹ 3a4053039d94 Waiting                                                                                                                                                                            0.2s
   ⠹ c2bf5e631951 Waiting                                                                                                                                                                            0.2s
 ⠿ kibana Error                                                                                                                                                                                      0.8s
no matching manifest for linux/arm64/v8 in the manifest list entries
```

bskinner commented:
> This looks awesome! Has anyone gotten this working on an M1 mac? When I try I get
> ...

Huh- that's curious. The images have been running on a ARM64 system (specifically a Raspberry Pi), so it's not immediately clear to me why Docker for macOS is not picking up the arm images, as both ES and Kibana have `linux/arm64` images. I haven't yet tested in true Apple Silicon as I do not have access to a Broad-supplied Apple Silicon Mac

~Have you tried explicitly adding the `--platform aarch64` flag?~ _(edit: looks like `docker compose` does not support this flag)_

bskinner committed:
Configuration cleanup to match `dev`'s structure

bskinner committed:
Allow 127.0.0.1 for CORS purposes

bskinner committed:
Bugfix - result should be inverted

bskinner commented:
Did some investigating- complete arm64 support is not going to be possible with our current dependencies. Elastic did not start offering `linux/arm64` builds until version `7.8.0`. DSS and DSM are currently targeting the `6.8.x` series of Elasticsearch releases, and engineering work is going to be needed to update to the 7.x releases or higher due to API changes.

devonbush commented:
Ah, bummer that this won't be able to work out-of-the-box for m1s for now.  Thanks for investigating!

bskinner committed:
DSS & DSM target ES 6.8.x

bskinner committed:
Bump up to the latest patch release of the ES client

bskinner committed:
DSS: Adds new Pub/Sub keys & documentation

* `pubsub.host`: the authority of the target Pub/Sub Emulator service
* `pubsub.housekeepingTasksTopic`: The Pub/Sub Topic to use for task messaging to Housekeeping
* `pubsub.pubSubTasksTopic`: The Pub/Sub Topic to use for task messaging to DSS
* `fileUploads.scanResultTopic`: The Pub/Sub Topic to listen on for results from the malware scanning service

bskinner committed:
Refactor of PubSubConnectionManager

* Use java.lang.ref.Cleaner for handling closing of resources. Removes the need for static cleanup of resources.
* Improves integration with the PubSub Emulator tool
* Streamline get/create behaviors for Topics & Subscriptions, and minimize the amount of error handling that's internally occurring.
* Updates for recent API improvements

bskinner committed:
Test cleanup

* Move globally static behavior into an interned instance
* Adds error handling when initializing Pub/Sub resources
* Create required topics & subscriptions on-demand when running tests.

bskinner committed:
Publicly expose if the emulator is in use

bskinner committed:
Keep the transaction window as small as possible

bskinner committed:
Connection manager cleanup

* Be sure to call awaitTimeout() after the shutdown request
* Move creating a Publisher builder to a separate method

bskinner committed:
Create necessary Pub/Sub resources at start() time

bskinner committed:
Remove the static config

This fixes a bug where multiple DSMConfig instances will overwrite the Config used by all live instances

bskinner committed:
Creates an emulator-safe Publisher

bskinner commented:
Closed due to inactivity.
